src/go/rpk/pkg/config/manager.go

```diff
@@ -368,13 +394,52 @@ func recover(fs afero.Fs, backup, path string, err error) error {
 }
 
 func unmarshal(v *viper.Viper) (*Config, error) {
-	conf := &Config{}
-	err := v.Unmarshal(conf)
+	result := &Config{}
+	decoderConfig := mapstructure.DecoderConfig{
+		Result:	result,
+		// Sometimes viper will save int values as strings (i.e.
+		// through BindPFlag) so we have to allow mapstructure
+		// to cast them.
+		WeaklyTypedInput:	true,
+		DecodeHook: mapstructure.ComposeDecodeHookFunc(
+			v21_1_4MapToNamedSocketAddressSlice,
+		),
+	}
+	decoder, err := mapstructure.NewDecoder(&decoderConfig)
 	if err != nil {
 		return nil, err
 	}
-	conf.ConfigFile, err = absPath(v.ConfigFileUsed())
-	return conf, err
+	err = decoder.Decode(v.AllSettings())
+	if err != nil {
+		return nil, err
+	}
+	result.ConfigFile, err = absPath(v.ConfigFileUsed())
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Redpanda version < 21.1.4 only supported a single anonymous listener and a
+// single anonymous advertised address. This custom decode function translates
+// a single SocketAddress-equivalent map[string]interface{} into a
+// []NamedSocketAddress.
+func v21_1_4MapToNamedSocketAddressSlice(
```

Comment:
Not related to this PR.

Why this function doesn't work in our initContainer? @dimitriscruz

Reply:
Not quite sure what you mean. Are you asking.. why is it that rpk doesn't work with the old (single address) configuration schema given that it uses this conversion function?

Reply:
Yes, why rpk doesn't work with old (single address)? Maybe I'm missing something.